src/resample2d.js

```diff
@@ -0,0 +1,147 @@
+        let ih = (oh + 0.5) / scaleHeight - 0.5;
+        if (ih < 0) {
+          ih = 0;
+        } else if (ih > srcHeight - 1) {
+          ih = srcHeight - 1;
+        }
+        for (let ow = 0; ow < dstWidth; ow++) {
+          let iw = (ow + 0.5) / scaleWidth - 0.5;
+          if (iw < 0) {
+            iw = 0;
+          } else if (iw > srcWidth - 1) {
+            iw = srcWidth - 1;
+          }
+          const outputValue = src.getValueByLocation(
+              [ob, Math.ceil(ih - 0.5), Math.ceil(iw - 0.5), oc]);
+          dst.setValueByLocation([ob, oh, ow, oc], outputValue);
+        }
+      }
+    }
+  }
+  return dst;
+}
+
+/**
+ * Resample the tensor values from the source to the destination spatial dimensions according to
+ * the scaling factors using Bilinear interpolation.
+ * Refer to https://en.wikipedia.org/wiki/Bilinear_interpolation
+ * @param {Tensor} src
+ * @param {Tensor} dst
+ * @param {Number} dstHeight
+ * @param {Number} dstWidth
+ * @param {Number} scaleHeight
+ * @param {Number} scaleWidth
+ * @return {Tensor}
+ */
+function linear(src, dst, dstHeight, dstWidth, scaleHeight, scaleWidth) {
+  const [srcBatches, srcHeight, srcWidth, srcChannels] = src.shape;
+  for (let ob = 0; ob < srcBatches; ob++) {
+    for (let oc = 0; oc < srcChannels; oc++) {
+      for (let oh = 0; oh < dstHeight; oh++) {
+        let ih = (oh + 0.5) / scaleHeight - 0.5;
+        if (ih < 0) {
+          ih = 0;
+        } else if (ih > srcHeight - 1) {
+          ih = srcHeight - 1;
+        }
+        const ihLower = Math.floor(ih);
+        const ihUpper = Math.ceil(ih);
+        const u = ih - ihLower;
+        for (let ow = 0; ow < dstWidth; ow++) {
+          let iw = (ow + 0.5) / scaleWidth - 0.5;
+          if (iw < 0) {
+            iw = 0;
+          } else if (iw > srcWidth - 1) {
+            iw = srcWidth - 1;
+          }
+          const iwLower = Math.floor(iw);
+          const iwUpper = Math.ceil(iw);
+          const v = iw - iwLower;
+          const outputValue =
+              (1 - u) * (1 - v) * src.getValueByLocation([ob, ihLower, iwLower, oc]) +
+              (1 - u) * v * src.getValueByLocation([ob, ihLower, iwUpper, oc]) +
+              u * (1 - v) * src.getValueByLocation([ob, ihUpper, iwLower, oc]) +
+              u * v * src.getValueByLocation([ob, ihUpper, iwUpper, oc]);
+          dst.setValueByLocation([ob, oh, ow, oc], outputValue);
+        }
+      }
+    }
+  }
+  return dst;
+}
+
+const interpolationFunctions = {
+  'nearest-neighbor': nearestNeighbor,
+  'linear': linear,
+};
+
+/**
+ * Resample the tensor values from the source to the destination spatial dimensions according to
+ * the scaling factors.
+ * @param {Tensor} input
+ * @param {MLResample2dOptions} options
+ * @return {Tensor}
+ */
+export function resample2d(
+    input,
+    {
+      mode = 'nearest-neighbor',
+      scales = [1.0, 1.0],
+      sizes,
+      axes = [2, 3],
+    } = {}) {
```

Comment:
(minor) It's kinda hard to follow these `} = {}) {`. Line break somewhere in there to aid readability?

Reply:
Updated, thanks.